### PR TITLE
[APPC-1634] Pay as a guest - Request Wallet/Entry Screen

### PR DESCRIPTION
**Pull request by JDSM01**
**What does this PR do?**

Implementation of the wallet generation for pay as a guest
Implementation of the entry screen for pay as a guest

**Database changed?**

No

**Where should the reviewer start?**

PaymentMethodsFragment.java

**How should this be manually tested?**

Check if the wallet is created and check if when you request with same id the wallet is the same
Enter the screen and test the flow

**What are the relevant tickets?**

https://aptoide.atlassian.net/browse/APPC-1634

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+          .putLong(MAX_BONUS_TTL_SECONDS_KEY, System.currentTimeMillis() / 1000)
+          .apply();
+    }
+  }
+
+  public boolean hasSavedBonus(long timeInMillis) {
+    if (sharedPreferences.contains(MAX_BONUS_KEY)) {
+      if (sharedPreferences.contains(MAX_BONUS_TTL_SECONDS_KEY)) {
+        return (timeInMillis / 1000 - sharedPreferences.getLong(MAX_BONUS_TTL_SECONDS_KEY, 0))
+            < 86400 * 30;
```
Comment (nunomonteiro) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java]: This should have a comment stating how long the bonus lasts
I guess it has something to do with 24h (86400 / 3600 ) but I have no idea why it is being multiplied by 30 afterwards. 6 months after no one will remember what 30 means?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  public void setMaxBonus(int maxBonus) {
+    if (maxBonus > 0) {
+      sharedPreferences.edit()
+          .putInt(MAX_BONUS_KEY, maxBonus)
+          .apply();
+      sharedPreferences.edit()
+          .putLong(MAX_BONUS_TTL_SECONDS_KEY, System.currentTimeMillis() / 1000)
+          .apply();
+    }
```
Comment (nunomonteiro) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java:43]: shouldn't it log error stating an invalid value is being used on else?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AndroidBillingMapper.java [hunk 3]
```diff
@@ -198,6 +198,55 @@ public static LaunchBillingFlowResult mapBundleToHashMapGetIntent(Bundle bundle)
...
+          String price = priceObj.getJSONObject("fiat")
+              .getString("value");
+          long priceAmountMicros = getFiatAmountInMicros(priceObj.getJSONObject("fiat"));
+          String priceCurrencyCode = getFiatCurrencyCode(priceObj.getJSONObject("fiat"));
+          if (priceObj.has("base") && priceObj.getString("base")
+              .equalsIgnoreCase(APPC)) {
+            price = getAppcPrice(priceObj);
+            priceAmountMicros = getAppcAmountInMicros(priceObj);
+            priceCurrencyCode = APPC;
+          }
```
Comment (nunomonteiro) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AndroidBillingMapper.java:224]: shouldn't the else clauses be implemented to deal with possible errors?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        + packageName;
+  }
+
+  public PackageManager getPackageManager() {
+    return packageManager;
+  }
+
+  public Intent getWalletInstallationIntent() {
+    final Intent cafeBazaarIntent = buildBrowserIntent(CAFE_BAZAAR_APP_URL);
+    if (false) { //WalletUtils.isCafeBazaarWalletAvailable()
```
Comment (nunomonteiro) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java]: Is this right? Should provide a comment stating why it's like this. ( I vaguely remember the rules of the game changing mid task, but it is important to note it here so that the next person picking this up understand the logic behind)

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      return redirectToRemainingStores(storeUrl);
+    }
+  }
+
+  private Intent cafeBazaarFlow(Intent cafeBazaarIntent, String storeUrl) {
+    if (WalletUtils.isAppInstalled(BuildConfig.CAFE_BAZAAR_PACKAGE_NAME, getPackageManager())
+        && isAbleToRedirect(cafeBazaarIntent)) {
+      cafeBazaarIntent.setPackage(BuildConfig.CAFE_BAZAAR_PACKAGE_NAME);
+      return cafeBazaarIntent;
+    } else if (false) { // CafeBazaarUtils.userFromIran(getUserCountry(getApplicationContext()))
```
Comment (nunomonteiro) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java]: Same as above:

Is this right? Should provide a comment stating why it's like this. ( I vaguely remember the rules of the game changing mid task, but it is important to note it here so that the next person picking this up understand the logic behind)

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/WalletInteract.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+public class WalletInteract {
+
+  private static int ID_LENGTH = 40;
+  private WalletRepository walletRepository;
+  private SharedPreferencesRepository sharedPreferencesRepository;
+
+  public WalletInteract(SharedPreferencesRepository sharedPreferencesRepository) {
+
+    this.walletRepository = new WalletRepository(new BdsService(BuildConfig.BACKEND_BASE),
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/WalletInteract.java]: why don't you inject the wallet repository?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java (hunk 1 above)
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java]: did you consider injecting the TTL value?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  public static String CREDIT_CARD_RADIO = "credit_card";
+  public static String PAYPAL_RADIO = "paypal";
+  public static String INSTALL_RADIO = "install";
+  private static String SELECTED_RADIO_KEY = "selected_radio";
+  private IabView iabView;
+  private BuyItemProperties buyItemProperties;
+  private PaymentMethodsPresenter paymentMethodsPresenter;
+  private PaymentMethodsFragmentLayout layout;
+  private String selectedRadioButton;
+  private SkuDetailsModel skuDetailsModel;
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java:41]: field not used

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+  private WalletGenerationModel walletGenerationModel;
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java:42]: field not used

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/models/WalletGenerationModel.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package com.appcoins.sdk.billing.models;
+
+public class WalletGenerationModel {
+
+  private final String walletAddress;
+  private final String ewt;
+  private boolean error;
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/models/WalletGenerationModel.java]: can you make error final?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/GamificationInteract.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    bdsService = new BdsService(BuildConfig.BACKEND_BASE);
+  }
+
+  public void loadMaxBonus(final MaxBonusListener maxBonusListener) {
+    if (sharedPreferencesRepository.hasSavedBonus(System.currentTimeMillis())) {
+      maxBonusListener.onBonusReceived(sharedPreferencesRepository.getMaxBonus());
+    } else {
+      ServiceResponseListener serviceResponseListener = new ServiceResponseListener() {
+        @Override public void onResponseReceived(RequestResponse requestResponse) {
+          GamificationMapper gamificationMapper = new GamificationMapper();
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/GamificationInteract.java]: why are you creating a new instance of GamificationMapper inside the listener?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/GamificationInteract.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import com.appcoins.billing.sdk.BuildConfig;
+import com.appcoins.sdk.billing.SharedPreferencesRepository;
+import com.appcoins.sdk.billing.service.BdsService;
+import com.appcoins.sdk.billing.service.RequestResponse;
+import com.appcoins.sdk.billing.service.ServiceResponseListener;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+class GamificationInteract {
+  private SharedPreferencesRepository sharedPreferencesRepository;
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/GamificationInteract.java]: can you set this field as final?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+import com.appcoins.sdk.billing.helpers.TranslationsXmlParser;
+import com.appcoins.sdk.billing.helpers.Utils;
+import java.util.Locale;
+
+import static com.appcoins.sdk.billing.helpers.Utils.RESPONSE_CODE;
+
+public class IabActivity extends Activity implements IabView {
+
+  private final static String TRANSLATIONS = "translations";
+  private AppcoinsBillingStubHelper appcoinsBillingStubHelper;
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: not used

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+  private BuyItemProperties buyItemProperties;
+  private TranslationsModel translationsModel;
+
+  @SuppressLint("ResourceType") @Override protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+
+    //This log is necessary for the automatic test that validates the wallet installation dialog
+    Log.d("InstallDialog", "com.appcoins.sdk.billing.helpers.InstallDialogActivity started");
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java:32]: good usage of documentation xD

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+    appcoinsBillingStubHelper = AppcoinsBillingStubHelper.getInstance();
+    buyItemProperties = (BuyItemProperties) getIntent().getSerializableExtra(
+        AppcoinsBillingStubHelper.BUY_ITEM_PROPERTIES);
+
+    if (savedInstanceState != null) {
+      translationsModel = (TranslationsModel) savedInstanceState.get(TRANSLATIONS);
+    } else {
+      fetchTranslations();
+      PaymentMethodsFragment paymentMethodsFragment = new PaymentMethodsFragment();
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: create newInstance method to create the fragment

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    if (translationsModel == null || !translationsModel.getLanguageCode()
+        .equalsIgnoreCase(locale.getLanguage()) || !translationsModel.getCountryCode()
+        .equalsIgnoreCase(locale.getCountry())) {
+      TranslationsXmlParser translationsParser = new TranslationsXmlParser(this);
+      translationsModel =
+          translationsParser.parseTranslationXml(locale.getLanguage(), locale.getCountry());
+    }
+  }
+
+  @SuppressLint("ResourceType")
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: why do you need to suppress lint here?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabView.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
+package com.appcoins.sdk.billing.payasguest;
+
+import android.content.Intent;
+import com.appcoins.sdk.billing.helpers.TranslationsModel;
+
+interface IabView {
+
+  TranslationsModel getTranslationsModel();
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabView.java:8]: not used

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+    Button positiveButton = layout.getPositiveButton();
+    RadioButton creditCardButton = layout.getCreditCardRadioButton();
+    RadioButton paypalButton = layout.getPaypalRadioButton();
+    RadioButton installRadioButton = layout.getInstallRadioButton();
+    RelativeLayout creditWrapper = layout.getCreditCardWrapperLayout();
+    RelativeLayout paypalWrapper = layout.getPaypalWrapperLayout();
+    RelativeLayout installWrapper = layout.getInstallWrapperLayout();
+    Button errorButton = layout.getErrorPositiveButton();
+    onRotation(savedInstanceState);
+    paymentMethodsPresenter.onCancelButtonClicked(cancelButton);
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java]: why do the presenter need views?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsInteract.java [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+  PaymentMethodsInteract(WalletInteract walletInteract, GamificationInteract gamificationInteract) {
+
+    this.walletInteract = walletInteract;
+    this.gamificationInteract = gamificationInteract;
+    this.paymentMethodsRepository =
+        new PaymentMethodsRepository(new BdsService(BuildConfig.HOST_WS));
+  }
+
+  String retrieveId() {
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsInteract.java]: should we use a more specific name?like retrieveWalletId or retriveUserId

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsPresenter.java [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+      paymentMethodsInteract.requestSkuDetails(buyItemProperties, listener);
+    } else {
+      fragmentView.showPaymentView();
+    }
+  }
+
+  private void loadPaymentsAvailable(String fiatPrice, String fiatCurrency) {
+    PaymentMethodsListener paymentMethodsListener = new PaymentMethodsListener() {
+      @Override public void onResponse(PaymentMethodsModel paymentMethodsModel) {
+        if (!paymentMethodsModel.hasError()) {
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsPresenter.java:91]: what if it has errors?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsView.java [hunk 20]
```diff
@@ -0,0 +1,30 @@
+package com.appcoins.sdk.billing.payasguest;
+
+import android.content.Intent;
+import com.appcoins.sdk.billing.models.WalletGenerationModel;
+
+interface PaymentMethodsView {
+  void setSkuInformation(SkuDetailsModel skuDetailsModel);
+
+  void showError();
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsView.java:9]: not used

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/SingleSkuDetailsAsync.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+import android.os.AsyncTask;
+import com.appcoins.billing.sdk.BuildConfig;
+import com.appcoins.sdk.billing.BuyItemProperties;
+import com.appcoins.sdk.billing.SkuDetails;
+import com.appcoins.sdk.billing.WSServiceController;
+import com.appcoins.sdk.billing.helpers.AndroidBillingMapper;
+import java.util.ArrayList;
+import java.util.List;
+
+class SingleSkuDetailsAsync extends AsyncTask<Object, Object, SkuDetails> {
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/SingleSkuDetailsAsync.java:12]: AsyncTasks are now deprecated!

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java [hunk 22]
```diff
@@ -111,6 +111,10 @@ public void makeRequest(String endPoint, String httpMethod, List<String> paths,
     AsyncTask asyncTask =
         new ServiceAsyncTask(this, baseUrl, endPoint, httpMethod, paths, queries, body,
             serviceResponseListener);
-    asyncTask.execute();
+    if (async) {
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java]: why did you add async flag?
is there any usage of it as false?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java [hunk 23]
```diff
@@ -111,6 +111,10 @@ public void makeRequest(String endPoint, String httpMethod, List<String> paths,
     AsyncTask asyncTask =
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java]: since this is using generic types, you should use ServiceAsyncTask specific type instead of the generic AsyncTask

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/wallet/WalletRepository.java [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+  private WalletGenerationMapper walletGenerationMapper;
+
+  public WalletRepository(Service service, WalletGenerationMapper walletGenerationMapper) {
+
+    this.service = service;
+    this.walletGenerationMapper = walletGenerationMapper;
+  }
+
+  public void requestWallet(String id, final WalletInteractListener walletInteractListener) {
+    Map<String, String> queries = new LinkedHashMap<>();
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/wallet/WalletRepository.java]: why use LinkedHashMap?

**Review by Trinkes — CHANGES_REQUESTED** [review]

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java (hunk 7 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java:41]: It will be used in the next ticket

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java (hunk 8 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsFragment.java:42]: Same as above. Will be used in the next ticket

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java (hunk 12 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: I was gonna use it on the next ticket, but I already remove it there. I'll remove it to

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabView.java (hunk 16 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabView.java:8]: Will be used in the future. Translation are not yet implemented in this ticket, but I wanted to inject the dependency already

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java (hunk 1 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java]: Yes, but without dagger it seems a bit odd having the fragment defining the TTL value. I think it should be extracted to a variable tho.

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java (hunk 22 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java]: As of now this variable will never be false, but since this is suppose to be a generic class, I'll leave here anyway. Changed the name tho, as of now it does not portrait what it intended (parallelism)

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java (hunk 1 above)
Comment (Trinkes) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java]: The way this class is used shouldn't impact the way it's organized. IMO, this class shouldn't define the TTL value. In the optimal solution, this should be defined application wide as a global setting.
If you don't want to define the TTL value on fragment, you can create a factory to help you create it.

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsInteract.java [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+  private final PaymentMethodsRepository paymentMethodsRepository;
+  private WalletInteract walletInteract;
+  private GamificationInteract gamificationInteract;
+
+  PaymentMethodsInteract(WalletInteract walletInteract, GamificationInteract gamificationInteract) {
+
+    this.walletInteract = walletInteract;
+    this.gamificationInteract = gamificationInteract;
+    this.paymentMethodsRepository =
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/PaymentMethodsInteract.java]: why don't you inject paymentMethodsRepository?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/Service.java [hunk 26]
```diff
@@ -14,6 +14,6 @@
    * @param serviceResponseListener Listener in which the response will be sent.
    */
   void makeRequest(String endPoint, String httpMethod, List<String> paths,
-      Map<String, String> queries, Map<String, Object> body,
+      Map<String, String> queries, Map<String, Object> body, boolean async,
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/Service.java]: this is never false, is it really needed?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java [hunk 27]
```diff
@@ -100,17 +100,21 @@ private RequestResponse handleException(HttpURLConnection urlConnection,
   }
 
   public void makeRequest(String endPoint, String httpMethod, List<String> paths,
-      Map<String, String> queries, Map<String, Object> body,
+      Map<String, String> queries, Map<String, Object> body, boolean shouldParallelize,
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java]: if you're gonna leave shouldParallelize flag here, you should also change its name on interface

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java [hunk 28]
```diff
@@ -100,17 +100,21 @@ private RequestResponse handleException(HttpURLConnection urlConnection,
   }
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/service/BdsService.java]: we should be careful about     byte[] input = body.getBytes(StandardCharsets.UTF_8);, I think it will crash in api<19

**Review by Trinkes — CHANGES_REQUESTED** [review]

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java (hunk 1 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java]: Agree

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AndroidBillingMapper.java (hunk 3 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AndroidBillingMapper.java:224]: The errors are dealt in the presenter. If there's any problem it will be catched and an error Json (default) will be created, when it reaches the presenter, the presenter checks for the Object (skuDetails) and deal with it there

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java (hunk 4 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/WalletInstallationIntentBuilder.java]: I was waiting for the ticket to be merged to dev :p

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java (hunk 2 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java:43]: We won't use the value equal or below 0. Most of the time it will happen when there's no network and that is being dealt in the presenter class, so I don't think it will be necessary.

**Review by Trinkes — APPROVED** [review]

**Review by nunomonteiro — APPROVED** [review]
